storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java

```diff
@@ -125,12 +126,37 @@ public RemoteIndexCache(RemoteStorageManager remoteStorageManager, String logDir
      * @param remoteStorageManager RemoteStorageManager instance, to be used in fetching indexes.
      * @param logDir               log directory
      */
-    public RemoteIndexCache(int maxSize, RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
+    public RemoteIndexCache(long maxSize, RemoteStorageManager remoteStorageManager, String logDir) throws IOException {
         this.remoteStorageManager = remoteStorageManager;
         cacheDir = new File(logDir, DIR_NAME);
 
-        internalCache = Caffeine.newBuilder()
-                .maximumSize(maxSize)
+        internalCache = initEmptyCache(maxSize);
+        init();
+
+        // Start cleaner thread that will clean the expired entries.
+        cleanerThread = createCleanerThread();
+        cleanerThread.start();
+    }
+
+    public void resizeCacheSize(long remoteLogIndexFileCacheSize) {
+        lock.writeLock().lock();
+        try {
+            Cache<Uuid, Entry> newCache = initEmptyCache(remoteLogIndexFileCacheSize);
+            for (Map.Entry<Uuid, Entry> entry : internalCache.asMap().entrySet()) {
+                newCache.put(entry.getKey(), entry.getValue());
+            }
+            internalCache = newCache;
```

Comment:
This could get messy as the entries are now present in two caches for some time and old one may try to get rid of them before going out of scope.

I would suggest to keep an expectation that resized cache always starts as empty. We don't need to eagerly populate it since resize is a rare operation.

Please also add a comment here about the reason why we aren't populating the new cache. It will help future readers make sense of the decisions we made in this code change.

Reply:
updated